Authentication feature

AlexNGerman opened the pull request:
Application configs: #absolute-imports
Created default pages structure for Authentication feature
updated package depedencies
added default createUser mutation
added routes constans

Commit:
Application configs: #absolute-imports
Created default pages structure for Authentication feature
updated package depedencies
added default createUser mutation
added routes constans

VladWithCookies commented on jsconfig.json:
```diff
@@ -0,0 +1,6 @@
+    "baseUrl": "src"
+  },
+  "include": ["src"]
+}
```
Missing final new line

VladWithCookies commented on package.json:
```diff
@@ -3,13 +3,18 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@emotion/react": "^11.7.1",
+    "@emotion/styled": "^11.6.0",
```
Do you use it?

AlexNGerman replied:
I added this dependencies because of compile error

_Failed to compile.
./node_modules/@mui/styled-engine/index.js
Module not found: Can't resolve '@emotion/react' in '/Users/alexnafangermanhotmail.com/web/internship-boilerplate/node_modules/@mui/styled-engine'_

VladWithCookies commented on package.json:
```diff
@@ -3,13 +3,18 @@
     "@testing-library/react": "^11.1.0",
     "@testing-library/user-event": "^12.1.10",
     "date-fns": "^2.27.0",
+    "eslint": "^7.11.0",
```
Should be a dev dependency

VladWithCookies commented on package.json:
```diff
@@ -3,13 +3,18 @@
+    "@emotion/styled": "^11.6.0",
+    "@mui/icons-material": "^5.2.5",
     "@mui/material": "^5.2.2",
     "@testing-library/jest-dom": "^5.11.4",
     "@testing-library/react": "^11.1.0",
     "@testing-library/user-event": "^12.1.10",
```
Should be a dev dependency

AlexNGerman replied:
That's from original repo. Maybe we need to move it to dev dependency in original repo too?

VladWithCookies replied:
> That's from original repo. Maybe we need to move it to dev dependency in original repo too?

Sure

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,31 @@
+
+import { APP_ROUTES } from 'constants/routes';
+
+class AppWrapper extends Component {
```
Why not to use functional component?

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,31 @@
+            <Route exact path={APP_ROUTES.home} element={<Home/>} />
+            <Route exact path={APP_ROUTES.signIn} element={<SignIn/>} />
+            <Route exact path={APP_ROUTES.signUp} element={<SignUp/>} />
+
```
Needless new line

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,31 @@
+
+          </Routes>
+        </BrowserRouter>
+
```
Needless new line

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,31 @@
+  }
+}
+
+export default AppWrapper;
```
I guess it should be named `Router`

VladWithCookies commented on src/utils/theme.js:
```diff
@@ -0,0 +1,6 @@
+import { createTheme } from '@mui/material/styles';
+
+
```
Needless new line

VladWithCookies commented on src/mutations/CreateUser/createUser.js:
```diff
@@ -0,0 +1,17 @@
+      lastName
+    }
+  }
+`
```
Missing final new line

VladWithCookies commented on src/mutations/CreateUser/createUser.js:
```diff
@@ -0,0 +1,17 @@
+      id
+      email
+      firstName
+      lastName
```
Do you really need these fields in response?

VladWithCookies commented on src/mutations/CreateUser/createUser.js:
```diff
@@ -0,0 +1,17 @@
+import { gql } from 'urql';
+
+export const CREATE_USER_MUTATION = gql`
```
simply `CREATE_USER`

VladWithCookies commented on src/index.js:
```diff
@@ -4,13 +4,20 @@ import { createClient, Provider } from 'urql';
 
 import { API_URL } from './constants/api';
 import reportWebVitals from './reportWebVitals';
+import { ThemeProvider } from '@mui/material/styles';
```
It would be better to group it together with other 3rd party imports

AlexNGerman replied:
Do you mean move ThemeProvider wrapper to Router module ?

VladWithCookies replied:
I meant import sorting

VladWithCookies commented on src/index.js:
```diff
@@ -4,13 +4,20 @@ import { createClient, Provider } from 'urql';
 import reportWebVitals from './reportWebVitals';
+import { ThemeProvider } from '@mui/material/styles';
+import theme from './utils/theme';
+
```
Needless newline

VladWithCookies commented on src/index.js:
```diff
@@ -4,13 +4,20 @@ import { createClient, Provider } from 'urql';
     <Provider value={client}>
-      Hello there!
+      <ThemeProvider theme={theme}>
+        <AppWrapper />
```
Router?

VladWithCookies commented on src/features/Home/pages/Home/Home.js:
```diff
@@ -0,0 +1,13 @@
+  }
+}
+
+export default Home;
```
Missing final new line

VladWithCookies commented on src/features/Home/pages/Home/Home.js:
```diff
@@ -0,0 +1,13 @@
+import React, { Component } from 'react';
+import PageTempate from 'components/templates/PageTempate';
+
+export class Home extends Component {
```
Use only functional components

VladWithCookies commented on src/features/Authentication/pages/SignUp/SignUp.js:
```diff
@@ -0,0 +1,14 @@
+import SignUpForm from 'features/Authentication/organisms/SignUpForm';
+import PageTempate from "components/templates/PageTempate/pageTemplate";
+
+export class SignUp extends Component {
```
Use only functional components

VladWithCookies commented on src/features/Authentication/pages/SignUp/SignUp.js:
```diff
@@ -0,0 +1,14 @@
+  }
+}
+
+export default SignUp;
```
Missing final new line

VladWithCookies commented on src/features/Authentication/pages/SignIn/SignIn.js:
```diff
@@ -0,0 +1,14 @@
+  }
+}
+
+export default SignIn;
```
Missing Final new line

VladWithCookies commented on src/features/Authentication/pages/SignIn/SignIn.js:
```diff
@@ -0,0 +1,14 @@
+
+  render() {
+    return (
+      <PageTempate children={<SignInForm />}/>
```
```
<MainTempate>
  <SignInForm />
</MainTemplate>
```

VladWithCookies commented on src/features/Authentication/pages/SignIn/SignIn.js:
```diff
@@ -0,0 +1,14 @@
+import SignInForm from 'features/Authentication/organisms/SignInForm';
+
+export class SignIn extends Component {
+
```
Needless new line

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,21 @@
+import React, { Fragment, Component } from 'react';
+// import { useNavigate } from 'react-router-dom';
+// import Button from "@mui/material/Button";
+//
+// import CREATE_USER_MUTATION from 'mutations/CreateUser/createUser';
+// import { useMutation } from 'urql';
```
Needless comments

AlexNGerman replied:
this will be used later, commented for now

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,21 @@
+// import CREATE_USER_MUTATION from 'mutations/CreateUser/createUser';
+// import { useMutation } from 'urql';
+
+class SignUpForm extends Component {
```
Use functional components

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,21 @@
+// import { useMutation } from 'urql';
+
+class SignUpForm extends Component {
+
```
Needless new line

VladWithCookies replied:
Still needless

AlexNGerman replied:
@VladWithCookies ?? it's old code

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,21 @@
+    return (
+      <Fragment>
+        <div>
+          {'SignUp form'}
```
Needless fragment, div and curly braces and

AlexNGerman replied:
Why? It's just placeholder copy for now

VladWithCookies replied:
This code can be simplified to:
```
<div>SignUp form</div>
```

VladWithCookies commented on src/features/Authentication/organisms/SignInForm/SignInForm.js:
```diff
@@ -0,0 +1,21 @@
+// import Button from "@mui/material/Button";
+//
+// import CREATE_USER_MUTATION from 'mutations/CreateUser/createUser';
+// import { useMutation } from 'urql';
```
Needless comments

VladWithCookies commented on src/features/Authentication/organisms/SignInForm/SignInForm.js:
```diff
@@ -0,0 +1,21 @@
+// import { useMutation } from 'urql';
+
+class SignInForm extends Component {
+
```
Needless new line

VladWithCookies commented on src/features/Authentication/organisms/SignInForm/SignInForm.js:
```diff
@@ -0,0 +1,21 @@
+// import CREATE_USER_MUTATION from 'mutations/CreateUser/createUser';
+// import { useMutation } from 'urql';
+
+class SignInForm extends Component {
```
Use functional components

VladWithCookies commented on src/features/Authentication/organisms/SignInForm/SignInForm.js:
```diff
@@ -0,0 +1,21 @@
+  render() {
+    return (
+      <Fragment>
+        <div>
```
Needless fragment, div and `{}`

AlexNGerman replied:
Why? It's just placeholder copy for now

VladWithCookies replied:
This code can be simplified

VladWithCookies commented on src/constants/routes.js:
```diff
@@ -0,0 +1,5 @@
+export const APP_ROUTES = {
+  home: '/',
+  signIn: '/sign-in',
+  signUp: '/sign-up'
```
Keys should be also in uppercase

VladWithCookies commented on src/components/templates/PageTempate/pageTemplate.js:
```diff
@@ -0,0 +1,44 @@
+                align="center"
+                color="text.primary"
+                gutterBottom
+              >{'TODO List App'}
```
Useless braces

VladWithCookies commented on src/components/templates/PageTempate/pageTemplate.js:
```diff
@@ -0,0 +1,44 @@
+
+import AppHeader from "../../organisms/AppHeader";
+
+class PageTemplate extends Component {
```
Use functional component

VladWithCookies commented on src/components/templates/PageTempate/pageTemplate.js:
```diff
@@ -0,0 +1,44 @@
+import React, {Component, Fragment} from 'react';
+import CssBaseline from "@mui/material/CssBaseline";
+import Typography from "@mui/material/Typography";
+import Box from "@mui/material/Box";
+import Container from "@mui/material/Container";
+
+import AppHeader from "../../organisms/AppHeader";
```
Use consistent quotes

VladWithCookies replied:
Do not use relative paths

VladWithCookies commented on src/components/templates/PageTempate/pageTemplate.js:
```diff
@@ -0,0 +1,44 @@
+import React, {Component, Fragment} from 'react';
```
pageTemplate.js -> MainTemplate.js

VladWithCookies commented on src/components/organisms/AppHeader/AppHeader.js:
```diff
@@ -0,0 +1,32 @@
+import Typography from "@mui/material/Typography";
+import Button from "@mui/material/Button";
+
+class AppHeader extends Component {
```
Use functional component

VladWithCookies commented on src/components/organisms/AppHeader/AppHeader.js:
```diff
@@ -0,0 +1,32 @@
+import Button from "@mui/material/Button";
+
+class AppHeader extends Component {
+
```
Needless new line

VladWithCookies replied:
Still here )

Commit:
resolve issues after review #1

Commit:
added "formik-mui"

Authentication feature:
- Sign in form

VladWithCookies commented on src/components/templates/MainTemplate/MainTemplate.js:
```diff
@@ -0,0 +1,22 @@
+import Container from "@mui/material/Container";
+import AppHeader from "components/organisms/AppHeader";
+
+class MainTemplate extends Component {
```
Use functional component

VladWithCookies commented on src/features/Authentication/organisms/SignInForm/SignInForm.js:
```diff
@@ -0,0 +1,11 @@
+
+const SignInForm = () => {
+  return (
+    <div>
+      'SignIn form'
+    </div>
```
Can be simplified to:
```
<div>
  SignIn form
</div>
```

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,30 @@
+import { APP_ROUTES } from 'constants/routes';
+
+const Router = () => {
+
```
Needless new line

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,30 @@
+
+const Router = () => {
+
+  return (
```
Needless return

AlexNGerman replied:
What do you mean? How it should be?

VladWithCookies replied:
```
const Router = () => (
  ...
);
```

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+
+const SIGNUP_SCHEMA = Yup.object().shape({
+  firstName: Yup.string()
+    .min(2, 'Too Short!')
+    .max(70, 'Too Long!')
```
2 and 70 can be extracted to the constants

VladWithCookies replied:
Error messages are not accessible. You need to specify required length in the error message.

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+  password: Yup.string()
+    .required("Please enter your password")
+    .matches(
+      /^.*(?=.{10,})((?=.*[!@#$%^&*()\-_=+{};:,<.>]){1})(?=.*\d)((?=.*[a-z]){1})((?=.*[A-Z]){1}).*$/,
```
Can be extracted to the constant and reused

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+  const onSubmit = async (values, { setSubmitting }, {error}) => {
+    await createUser(values).then(result => {
+      if (result.error) {
+        console.error('Oh no!', result.error.message);
```
Is not very useful approach for error handling

AlexNGerman replied:
How it can be changed?

VladWithCookies replied:
You need to parse error message and display it near the form

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+        initialValues={INITIAL_VALUES}
+        validationSchema={SIGNUP_SCHEMA}
+        onSubmit={(values, setSubmitting) => {
+          onSubmit(values, setSubmitting, {error});
```
It would be better to take error from response instead

VladWithCookies replied:
do onSubmit={onSubmit} instead

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+                    <Field component={TextField} name="firstName" fullWidth label="First name" placeholder="First name" variant="outlined" disabled={isSubmitting}/>
+                  </Grid>
+                  <Grid item xs={12}>
+                    <Field component={TextField} name="lastName" fullWidth label="Last name" placeholder="Last name" variant="outlined" disabled={isSubmitting}/>
```
It is unnecessary to disable inputs while submitting

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+                      sx={{ mt: 3, mb: 2 }}
+                      disabled={isSubmitting}
+                    >
+                      Sign In {isSubmitting && fetching && <CircularProgress color="inherit" /> }
```
fetching would be enough

AlexNGerman replied:
disabled={isSubmitting} and for CircularProgress ?

VladWithCookies replied:
`isSubmitting && fetching && <CircularProgress... ` -> `fetching && <CircularProgress... `

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,125 @@
+
+export default SignUpForm;
+
+
```
Needless new line

Commit:
resolve issues after review #2

Commit:
new lines deleted

Commit:
resolve issues after review #3

Commit:
set state for success Alert

Commit:
resolve issues after review #4

Commit:
first SignUp test
forms fields test constants

Commit:
resolve issues after review #5
remove useState hook

Commit:
upd test(not worked)

Commit:
removed needless fields from mutation responce
updated MainTemplate style
added wonka package
updated dependencies ver.
signup test - draft

Commit:
remove unused init field
added msw package

VladWithCookies commented on src/router.js:
```diff
@@ -0,0 +1,26 @@
+import { APP_ROUTES } from 'constants/routes';
+
+const Router = () => (
+  <Fragment>
```
Fragment is useless here

VladWithCookies commented on src/mutations/SignInUser/signInUser.js:
```diff
@@ -0,0 +1,15 @@
+    signInUser(email: $email, password: $password) {
+      token
+      user {
+        email
+        firstName
+        id
+        lastName
```
Do you really need email, firstName and lastName in response?

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/__tests__/SignUpForm.test.js:
```diff
@@ -0,0 +1,54 @@
+    });
+  });
+
+})
```
Missing final new line

VladWithCookies commented on src/features/Authentication/organisms/SignInForm/SignInForm.js:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+const SignInForm = () => {
+  return (
+    <div>
+      SignIn form
+    </div>
+  );
+}
```
```
const SignInForm = () => (
  <div>
    SignIn form
  </div>
);
```

VladWithCookies commented on src/constants/formTest.constants.js:
```diff
@@ -0,0 +1,6 @@
+export const SIGNUP_FIELDS = {
+  FIRST_NAME: 'John',
+  LAST_NAME: 'Dee',
+  EMAIL: 'john.dee@someemail.com',
+  PASSWORD: 'Alex_12345'
+};
```
Not sure we need this constant at all

AlexNGerman replied:
@VladWithCookies why not? on Sign In form tests we will need email & password

AlexNGerman replied:
@VladWithCookies i just renamed it to - AUTH_FIELDS

VladWithCookies replied:
It is just a mock data for the tests, it is not very useful to store it in the constant

VladWithCookies commented on src/constants/auth.js:
```diff
@@ -0,0 +1,6 @@
+export const AUTH_VALIDATION_SETS = {
+  MIN_LENGTH: 2,
+  MAX_LENGTH: 70,
+  PASSWORD_LENGTH: 10,
+  PASSWORD_MATCH_STRING: /^.*(?=.{10,})((?=.*[!@#$%^&*()\-_=+{};:,<.>]){1})(?=.*\d)((?=.*[a-z]){1})((?=.*[A-Z]){1}).*$/
```
It would be better to keep them as separate constants

VladWithCookies commented on src/components/templates/MainTemplate/MainTemplate.js:
```diff
@@ -0,0 +1,20 @@
+import Container from "@mui/material/Container";
+import AppHeader from "components/organisms/AppHeader";
+
+const MainTemplate = (props) =>{
+  return (
+    <Fragment>
+      <CssBaseline />
+      <AppHeader />
+      <main>
+        <Container maxWidth="md" align="center">
+          {props.children}
+        </Container>
+      </main>
+    </Fragment>
+  );
+}
```
```js
  const MainTemplate = ({ children }) => (
    <Fragment>
      <CssBaseline />
      <AppHeader />
      <main>
        <Container maxWidth="md" align="center">
          {children}
        </Container>
      </main>
    </Fragment>
  );
```

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,129 @@
+                      sx={{ mt: 3, mb: 2 }}
+                      disabled={isSubmitting}
+                    >
+                      {fetching ? <CircularProgress color="inherit" /> : "Sign Up"}
```
Can be moved inside the button conponent

AlexNGerman replied:
@VladWithCookies what do you mean?

VladWithCookies replied:
@AlexNGerman it would be nice to create a new molecule called SubmitButton and manage displaying of spinner inside this component

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,129 @@
+                Sign Up
+              </Typography>
+
+              {(data && data.createUser !== null) &&
```
Use state to track this

AlexNGerman replied:
@VladWithCookies will use a hook useState above ?

or something like: `const success = data && data.createUser !== null;`  ?

VladWithCookies replied:
@AlexNGerman lets discuss this

Commit:
resolve issues after review #5

Commit:
resolve issues after review #6

Commit:
fix data-testid

Commit:
AUTH_FIELDS

Commit:
resolve issues after review #7

Commit:
removed constants fields

VladWithCookies commented on src/constants/auth.js:
```diff
@@ -0,0 +1,4 @@
+export const MIN_LENGTH = 2;
+export const MAX_LENGTH = 70;
+export const PASSWORD_LENGTH = 10;
+export const PASSWORD_MATCH_STRING = /^.*(?=.{10,})((?=.*[!@#$%^&*()\-_=+{};:,<.>]){1})(?=.*\d)((?=.*[a-z]){1})((?=.*[A-Z]){1}).*$/ ;
```
PASSWORD_REGEX

VladWithCookies commented on src/features/Authentication/molecules/SubmitButton/SubmitButton.js:
```diff
@@ -0,0 +1,16 @@
+import { Button, CircularProgress } from '@mui/material';
+
+const SubmitButton = ({children, loading }) => (
```
Missed space

VladWithCookies commented on src/features/Authentication/molecules/SubmitButton/SubmitButton.js:
```diff
@@ -0,0 +1,16 @@
+  </Button>
+);
+
+export default SubmitButton;
```
Missing final new line

VladWithCookies commented on src/features/Authentication/organisms/SignUpForm/SignUpForm.js:
```diff
@@ -0,0 +1,151 @@
+import { TextField } from 'formik-mui';
+import Box from '@mui/material/Box';
+import Grid from '@mui/material/Grid';
+import SubmitButton from 'features/Authentication/molecules/SubmitButton';
```
Should be grouped with other our modules

Commit:
resolve issues after review #8
